### PR TITLE
Routing rule gsv2

### PR DESCRIPTION
**Pull request by zzxgzgz**
What does this PR do:

This PR implemented the routing rule update with GoalStateV2.

NOTE:
Please don't merge it until it is tested.

### CODE REVIEW

**Comment by zzxgzgz** [comment]
Merged code from upstream master.

**Comment by zzxgzgz** [comment]
Will create an issue to track the testing of this item.

File: src/ovs/aca_ovs_l3_programmer.cpp [hunk 1]
```diff
@@ -829,6 +829,107 @@ int ACA_OVS_L3_Programmer::create_or_update_router(RouterConfiguration &current_
...
+                  auto subnet_iterator = parsed_struct.subnet_states().find(subnet_id_of_fixed_ip);
+                  if(subnet_iterator != parsed_struct.subnet_states().end()){
+                    gw_mac = subnet_iterator->second.configuration().gateway().mac_address();
+                    dest_tunnel_id = subnet_iterator->second.configuration().tunnel_id();
+                    ACA_LOG_INFO("gw_mac: %s\n", gw_mac.c_str());
+                    ACA_LOG_INFO("dest_tunnel_id: %d\n", dest_tunnel_id);
+                  }else{
+                    ACA_LOG_INFO("Founding find subnet ID: [%s] in the goalstate", subnet_id_of_fixed_ip);
+                  }
+                  // for (auto each_subnet : parsed_struct.subnet_states()) {
```
Comment (lfu-ps) [inline src/ovs/aca_ovs_l3_programmer.cpp:869]: remove dead code please

File: src/ovs/aca_ovs_l3_programmer.cpp [hunk 2]
```diff
@@ -829,6 +829,107 @@ int ACA_OVS_L3_Programmer::create_or_update_router(RouterConfiguration &current_
...
+                auto current_fixed_ip = current_NeighborConfiguration1.fixed_ips(y);
+                string virtual_ip_address = current_fixed_ip.ip_address();
+                string virtual_mac_address =
+                        current_NeighborConfiguration1.mac_address();
+                string subnet_id_of_fixed_ip = current_fixed_ip.subnet_id();
+                string gw_mac;
+                uint dest_tunnel_id = 0;
+
+                if (strcmp(current_routing_rule.next_hop_ip().c_str(),
+                            current_fixed_ip.ip_address().c_str()) == 0) {
```
Comment (lfu-ps) [inline src/ovs/aca_ovs_l3_programmer.cpp:859]: indentation

File: src/ovs/aca_ovs_l3_programmer.cpp [hunk 3]
```diff
@@ -829,6 +829,107 @@ int ACA_OVS_L3_Programmer::create_or_update_router(RouterConfiguration &current_
             new_routing_rule_entry.next_hop_mac =
                     current_routing_rule.routing_rule_extra_info().next_hop_mac();
 
+            auto remote_host_ip = "";
+            ulong culminative_dataplane_programming_time = 0;
+            for (auto each_neighbor : parsed_struct.neighbor_states()) {
+              NeighborConfiguration current_NeighborConfiguration1 =
```
Comment (lfu-ps) [inline src/ovs/aca_ovs_l3_programmer.cpp:835]: why 'Configuration1' ?

File: src/ovs/aca_ovs_l3_programmer.cpp [hunk 4]
```diff
@@ -829,6 +829,107 @@ int ACA_OVS_L3_Programmer::create_or_update_router(RouterConfiguration &current_
...
+                string virtual_mac_address =
+                        current_NeighborConfiguration1.mac_address();
+                string subnet_id_of_fixed_ip = current_fixed_ip.subnet_id();
+                string gw_mac;
+                uint dest_tunnel_id = 0;
+
+                if (strcmp(current_routing_rule.next_hop_ip().c_str(),
+                            current_fixed_ip.ip_address().c_str()) == 0) {
+                  auto subnet_iterator = parsed_struct.subnet_states().find(subnet_id_of_fixed_ip);
+                  if(subnet_iterator != parsed_struct.subnet_states().end()){
```
Comment (lfu-ps) [inline src/ovs/aca_ovs_l3_programmer.cpp:861]: space between 'if' and '(', and before '{' please

File: src/ovs/aca_ovs_l3_programmer.cpp [hunk 5]
```diff
@@ -829,6 +829,107 @@ int ACA_OVS_L3_Programmer::create_or_update_router(RouterConfiguration &current_
...
+                    gw_mac = subnet_iterator->second.configuration().gateway().mac_address();
+                    dest_tunnel_id = subnet_iterator->second.configuration().tunnel_id();
+                    ACA_LOG_INFO("gw_mac: %s\n", gw_mac.c_str());
+                    ACA_LOG_INFO("dest_tunnel_id: %d\n", dest_tunnel_id);
+                  }else{
```
Comment (lfu-ps) [inline src/ovs/aca_ovs_l3_programmer.cpp:866]: space

File: src/ovs/aca_ovs_l3_programmer.cpp (hunk 3 above)
Comment (zzxgzgz) [reply src/ovs/aca_ovs_l3_programmer.cpp:835]: > why 'Configuration1' ?

Just following the variable names in the V1 version.

File: src/ovs/aca_ovs_l3_programmer.cpp (hunk 2 above)
Comment (zzxgzgz) [reply src/ovs/aca_ovs_l3_programmer.cpp:859]: fixed.

File: src/ovs/aca_ovs_l3_programmer.cpp (hunk 4 above)
Comment (zzxgzgz) [reply src/ovs/aca_ovs_l3_programmer.cpp:861]: added

File: src/ovs/aca_ovs_l3_programmer.cpp (hunk 5 above)
Comment (zzxgzgz) [reply src/ovs/aca_ovs_l3_programmer.cpp:866]: added

File: src/ovs/aca_ovs_l3_programmer.cpp (hunk 1 above)
Comment (zzxgzgz) [reply src/ovs/aca_ovs_l3_programmer.cpp:869]: removed.

**Review by lfu-ps — APPROVED** [review]

**Pull request merged**